### ISSUE TITLE
HttpServer metrics are broken since 2.1.5

### ISSUE THREAD

**Issue by Delorien84** [spring-projects/spring-boot#17103]
Hi,
 we have notice that when we upgrade to Spring Boot 2.1.5, the http server request metrics does not work properly. It only counts requests if the response has empty body. It does not matter if mime type is JSON or XML, does not matter of request type. Eg.: in the following simple example, `/jsonGet` request is never count.

```
@RestController
@RequestMapping(path = "/")
public class Controller {
  
    @GetMapping(path = "/jsonGet")
    public Mono<Map<String, String>> jsonGet() {
        return Mono.just(Map.of("key", "value"));
    }

    @GetMapping(path = "/jsonGetEmpty")
    public Mono<Map<String, String>> jsonGetEmpty() {
        return Mono.empty();
    }
}
```
Simple project, generated from https://start.spring.io/. Using Postman to call the endpoint.

I was able to truck down to `org.springframework.boot.actuate.metrics.web.reactive.server.MetricsWebFilter.filter()` where `doOnSuccess` callback is never called. Seems to be related with `Reactor Netty v0.8.8.RELEASE` performance improvements.

Upgrading to `Reactor Netty 0.8.9` and `Reactor Core v3.2.10` does not help. Upgrading to current snapshot of `Spring Boot 2.1.6` does not help. 

Working correctly in `Spring Boot 2.1.4`

**Comment by wilkinsona** [comment]
Thanks for the report and initial analysis. This looks to me like a Reactor Netty or WebFlux problem rather than something caused by Boot. I've asked them to take a look.

**Comment by bclozel** [comment]
I've reproduced the issue and:
* Spring Boot 2.1.5.RELEASE + Reactor Netty 0.8.8.RELEASE -> issue reproduced 
* Spring Boot 2.1.5.RELEASE + Reactor Netty 0.8.7.RELEASE -> issue reproduced
* Spring Boot 2.1.5.RELEASE + Reactor Netty 0.8.6.RELEASE -> working correctly

This only seems to happen when the controller returns value with a `Mono`, like (an empty Mono works fine):

```
@RestController
public class SampleController {

	@GetMapping(path = "/empty")
	public Mono<String> empty() {
		return Mono.empty();
	}

	@GetMapping("/text")
	public Mono<String> test() {
		return Mono.just(("TEXT"));
	}

}
```

I've locally instrumented the `MetricsWebFilter` and added a `log()` operator right after the `filter()` call here (line 76): 
https://github.com/spring-projects/spring-boot/blob/b5d19b798eca7a6400a97e80d42601e027789aa6/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/web/reactive/server/MetricsWebFilter.java#L73-L79

## With Reactor Netty 0.8.6.RELEASE

```
2019-06-11 12:48:05.158 DEBUG 18517 --- [ctor-http-nio-2] o.s.w.s.adapter.HttpWebHandlerAdapter    : [b15e6f08] HTTP GET "/text"
2019-06-11 12:48:08.617  INFO 18517 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : onSubscribe([Fuseable] MonoFlatMap.FlatMapMain)
2019-06-11 12:48:08.623  INFO 18517 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : request(unbounded)
2019-06-11 12:48:08.699 DEBUG 18517 --- [ctor-http-nio-2] s.w.r.r.m.a.RequestMappingHandlerMapping : [b15e6f08] Mapped to public reactor.core.publisher.Mono<java.lang.String> com.example.demo.SampleController.test()
2019-06-11 12:48:08.706 DEBUG 18517 --- [ctor-http-nio-2] o.s.w.r.r.m.a.ResponseBodyResultHandler  : Using 'text/plain;charset=UTF-8' given [*/*] and supported [text/plain;charset=UTF-8, text/event-stream, text/plain;charset=UTF-8, */*]
2019-06-11 12:48:08.706 DEBUG 18517 --- [ctor-http-nio-2] o.s.w.r.r.m.a.ResponseBodyResultHandler  : [b15e6f08] 0..1 [java.lang.String]
2019-06-11 12:48:08.709 DEBUG 18517 --- [ctor-http-nio-2] o.s.core.codec.CharSequenceEncoder       : [b15e6f08] Writing "TEXT"
2019-06-11 12:48:08.723  INFO 18517 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : onComplete()
2019-06-11 12:48:09.684 DEBUG 18517 --- [ctor-http-nio-2] o.s.w.s.adapter.HttpWebHandlerAdapter    : [b15e6f08] Completed 200 OK
```

## With Reactor Netty 0.8.7.RELEASE
```
2019-06-11 12:49:19.831 DEBUG 18528 --- [ctor-http-nio-2] o.s.w.s.adapter.HttpWebHandlerAdapter    : [bc8dd498] HTTP GET "/text"
2019-06-11 12:49:24.262  INFO 18528 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : onSubscribe([Fuseable] MonoFlatMap.FlatMapMain)
2019-06-11 12:49:24.269  INFO 18528 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : request(unbounded)
2019-06-11 12:49:24.340 DEBUG 18528 --- [ctor-http-nio-2] s.w.r.r.m.a.RequestMappingHandlerMapping : [bc8dd498] Mapped to public reactor.core.publisher.Mono<java.lang.String> com.example.demo.SampleController.test()
2019-06-11 12:49:24.349 DEBUG 18528 --- [ctor-http-nio-2] o.s.w.r.r.m.a.ResponseBodyResultHandler  : Using 'text/plain;charset=UTF-8' given [*/*] and supported [text/plain;charset=UTF-8, text/event-stream, text/plain;charset=UTF-8, */*]
2019-06-11 12:49:24.349 DEBUG 18528 --- [ctor-http-nio-2] o.s.w.r.r.m.a.ResponseBodyResultHandler  : [bc8dd498] 0..1 [java.lang.String]
2019-06-11 12:49:24.353 DEBUG 18528 --- [ctor-http-nio-2] o.s.core.codec.CharSequenceEncoder       : [bc8dd498] Writing "TEXT"
2019-06-11 12:49:24.377  INFO 18528 --- [ctor-http-nio-2] reactor.Mono.Defer.1                     : cancel()
```

For comparison, the empty case yields:

```
2019-06-11 12:51:43.384 DEBUG 18538 --- [ctor-http-nio-3] o.s.w.s.adapter.HttpWebHandlerAdapter    : [b8136b91] HTTP GET "/empty"
2019-06-11 12:51:43.384  INFO 18538 --- [ctor-http-nio-3] reactor.Mono.Defer.2                     : onSubscribe([Fuseable] MonoFlatMap.FlatMapMain)
2019-06-11 12:51:43.384  INFO 18538 --- [ctor-http-nio-3] reactor.Mono.Defer.2                     : request(unbounded)
2019-06-11 12:51:43.385 DEBUG 18538 --- [ctor-http-nio-3] s.w.r.r.m.a.RequestMappingHandlerMapping : [b8136b91] Mapped to public reactor.core.publisher.Mono<java.lang.String> com.example.demo.SampleController.empty()
2019-06-11 12:51:43.385 DEBUG 18538 --- [ctor-http-nio-3] o.s.w.r.r.m.a.ResponseBodyResultHandler  : Using 'text/plain;charset=UTF-8' given [*/*] and supported [text/plain;charset=UTF-8, text/event-stream, text/plain;charset=UTF-8, */*]
2019-06-11 12:51:43.385 DEBUG 18538 --- [ctor-http-nio-3] o.s.w.r.r.m.a.ResponseBodyResultHandler  : [b8136b91] 0..1 [java.lang.String]
2019-06-11 12:51:43.387  INFO 18538 --- [ctor-http-nio-3] reactor.Mono.Defer.2                     : onComplete()
2019-06-11 12:51:43.387 DEBUG 18538 --- [ctor-http-nio-3] o.s.w.s.adapter.HttpWebHandlerAdapter    : [b8136b91] Completed 200 OK
```

**Comment by ask4gilles** [comment]
@bclozel if it may help, this has been raised also here https://github.com/opentracing-contrib/java-spring-jaeger/issues/50 See my comment here https://github.com/opentracing-contrib/java-spring-jaeger/issues/50#issuecomment-500973278 : it looks like a change in reactor netty.

**Comment by duke-cliff** [comment]
Found similar issues on my end:
_doOnEach_ is never called after the SB 2.1.6 (from 2.1.0)upgrade.
```
 @Override
  public Mono<Void> filter(final ServerWebExchange exchange, final WebFilterChain chain) {
    return chain.filter(exchange).compose((call) -> filter(exchange, call));
  }

  private Publisher<Void> filter(ServerWebExchange exchange, Mono<Void> call) {

    return call.doOnEach(voidSignal -> {
      Context context = voidSignal.getContext();
      if (voidSignal.isOnComplete() || voidSignal.isOnError()) {
        logger.trace("Audit Interceptor Completed");

        Optional<AuditableContext> auditableContext = context.getOrEmpty(AUDITABLE_CONTEXT);
        if (auditableContext.isPresent()) {
          auditableContext.get().requestCompleted(exchange.getResponse());
        }
      }
    }).subscriberContext(context -> {
      AuditableContext auditableContext = auditProvider.prepare(exchange.getRequest());
      logger.trace("Audit Interceptor Started");
      auditableContext.requestStart();
      return context.put(AUDITABLE_CONTEXT, auditableContext);
    });
  }
```

**Comment by duke-cliff** [comment]
Also it works when tried downgrade reactornetty to 0.8.6. @wilkinsona @bclozel

**Comment by bclozel** [comment]
I'm closing this issue as a duplicate of:

* spring-projects/spring-framework#22952 in Spring Framework, which has been solved already for Spring Framework 5.1.8 and Spring Boot 2.1.6.RELEASE
* reactor/reactor-netty#741 which is the remaining part of this problem
